### ISSUE TITLE
Unexpected promise element from allSettled result when mixing promise and non promise elements in array argument.

### ISSUE THREAD

**Issue by Ranguna** [microsoft/TypeScript#46048]
# lib Update Request

## Configuration Check

My compilation *target* is `ES2019` and my *lib* is `ES2020.Promise`.

## Missing / Incorrect Definition
The definitions for `Promise.allSettled` don't seem to be correct for mix of promise and non promise values.

## Sample Code
```
let startingList = ["hello", "there"];

let resultingValues = await Promise.allSettled(
  startingList.map(string => string === "hello" ? 123 : Promise.resolve("abc"))
);

if(resultingValues[0].status === "fulfilled") console.log(resultingValues[0].value);
```
https://www.typescriptlang.org/play?target=7#code/DYUwLgBAzmCGBOYCWA7A5gGSTCBeCA2gEQAWIwwA9kQDQRFhnwhEC6A3AFCeiTNQBXYMnQA1WMAEgoeCLADusJJAAK8SgFtsIAHQTgAZXBhQAEwAUnCNDiJUmbGB0bYAB3Mx49vAD4bX9DxcfFJyKiIIAH4IAEYAJgBmCAAuCDVNbR1+SmAANxBzIlgAIwBjIgBKCs4Krk4kADNzfiERNHFJaQIABlYdGFgwARlgkIahBqQKEFNKiFLKFCgc3So0ZulW+w6pKB6+3IkpWqA

In this example code, the type of `resultingValues[0].value` is `123 | Promise<string>`, but I was expecting something like `123 | string` since all promise values should be resolved and all non promise values should be returned as is.

The type is inferred correctly by removing this definition from `lib/lib.es2020.promise.d.ts`:
```
allSettled<T extends readonly unknown[] | readonly [unknown]>(values: T):
        Promise<{ -readonly [P in keyof T]: PromiseSettledResult<T[P] extends PromiseLike<infer U> ? U : T[P]> }>;
```
But without this one, we'll lose nice inference for things like tuples.

It also works by forcing typescript to use the other definition by setting a non array type argument:
```
Promise.allSettled<Promise<string> | number>(
  startingList.map(string => string === "hello" ? 123 : Promise.resolve("abc"))
);
```
https://www.typescriptlang.org/play?target=7#code/DYUwLgBAzmCGBOYCWA7A5gGSTCBeCA2gEQAWIwwA9kQDQRFhnwhEC6A3AFCeiTNQBXYMnQA1WMAEgoeCLADusJJAAK8SgFtsIAHQTgAZXBhQAEwA8azdvMx4qNAD4IAHwgoBGgEYh4jgBScENBwiA5YMDoasAAO-nYOeM4J6Hi4+KTkVEQQAPwQAIwATADMEABcEFZaULr8lMAAbiD+RLBeAMZEAJTdnN1cnEgAZv78QiJo4pLSBAAMrDowsGACMukZw0LDSBQgpj0QHZQoUA26VGhj0hMO01JQ84uNElIDQA

## Documentation Link
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled

### Related issues:
Close but not really: https://github.com/microsoft/TypeScript/issues/42271

**Comment by andrewbranch** [comment]
Make sure to try the nightly build in the playground when you open bugs—this is already fixed and will ship with 4.5 🥳 

https://www.typescriptlang.org/play?target=7&ts=4.5.0-dev.20210929#code/IYZwngdgxgBAZgV2gFwJYHsLwBQEoYDeAUDDADYCmyMIywATmhAOYAyqtMAvDANoBEACwpky6fgBoY-ZMPoV+AXQDcREuSox5IBGSbMAasDIIKIbjGAB3YKmoAFeugC2HCgDpjZAMpVklABNsdVJaBn12WndnYAAHbFp6VBZuAD4aZCSUrhzpYVFxGAB+GABGACYAZhgALhhHFzd3bXQyADcKbH5gACMoflxcdVxVdVQ4bG1dfSMTM14ABkV3MOQEcxyefkQyOFRRCgCBmChMEFaPMWZJs2nkw2NTEEXltseKEaIAXyA

**Comment by Ranguna** [comment]
@andrewbranch oh man that's amazing!
So sorry about this one.
Thank you for the awesome work!
